File: scripts/src/display/CanvasRenderingContext2D.ts
//////////////////////////////////////////////////////////////////////////////////////
//
//  The MIT License (MIT)
//
//  Copyright (c) 2017-present, cyder.org
//  All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy of
//  this software and associated documentation files (the "Software"), to deal in the
//  Software without restriction, including without limitation the rights to use, copy,
//  modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
//  and to permit persons to whom the Software is furnished to do so, subject to the
//  following conditions:
//
//      The above copyright notice and this permission notice shall be included in all
//      copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
//  INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
//  PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
//  HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
//  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
//  SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//
//////////////////////////////////////////////////////////////////////////////////////

/**
 * The CanvasRenderingContext2D interface is used for drawing rectangles, text, images and other objects onto the render
 * object. It provides the 2D rendering context for the drawing surface of a render object.<br/>
 * To get an object of this interface, call getContext() on a render object, supplying "2d" as the contextType argument.
 */
interface CanvasRenderingContext2D {
    /**
     * A read-only back-reference to the associated render object.
     */
    readonly canvas:Canvas;
    /**
     * Alpha value that is applied to shapes and images before they are composited onto the canvas.
     * @default 1.0
     */
    globalAlpha:number;
    /**
     * Sets the type of compositing operation to apply when drawing new shapes, where type is a string identifying which
     * of the compositing or blending mode operations to use.
     * @default "source-over"
     */
    globalCompositeOperation:string;
    /**
     * Image smoothing mode; if disabled, images will not be smoothed if scaled.
     * @default false
     */
    imageSmoothingEnabled:boolean;
    /**
     * Draws an image onto the canvas.
     * @param image An image to draw into the context.
     * @param targetX The X coordinate in the destination canvas at which to place the top-left corner of the source image.
     * @param targetY The Y coordinate in the destination canvas at which to place the top-left corner of the source image.
     */
    drawImage(image:CanvasImageSource, targetX:number, targetY:number);
    /**
     * Draws an image onto the canvas.
     * @param image An image to draw into the context.
     * @param targetX The X coordinate in the destination canvas at which to place the top-left corner of the source image.
     * @param targetY The Y coordinate in the destination canvas at which to place the top-left corner of the source image.
     * @param targetWidth The width to draw the image in the destination canvas. This allows scaling of the drawn image.
     * If not specified, the image is not scaled in width when drawn.
     * @param targetHeight The height to draw the image in the destination canvas. This allows scaling of the drawn image.
     * If not specified, the image is not scaled in height when drawn.
     */
    drawImage(image:CanvasImageSource, targetX:number, targetY:number, targetWidth:number, targetHeight:number);
    /**
     * Draws an image onto the canvas.
     * @param image An image to draw into the context.
     * @param sourceX The X coordinate of the top left corner of the sub-rectangle of the source image to draw into the
     * destination context.
     * @param sourceY The Y coordinate of the top left corner of the sub-rectangle of the source image to draw into the
     * destination context.
     * @param sourceWidth The width of the sub-rectangle of the source image to draw into the destination context.
     * @param sourceHeight The height of the sub-rectangle of the source image to draw into the destination context.
     * @param targetX The X coordinate in the destination canvas at which to place the top-left corner of the source image.
     * @param targetY The Y coordinate in the destination canvas at which to place the top-left corner of the source image.
     * @param targetWidth The width to draw the image in the destination canvas. This allows scaling of the drawn image.
     * If not specified, the image is not scaled in width when drawn.
     * @param targetHeight The height to draw the image in the destination canvas. This allows scaling of the drawn image.
     * If not specified, the image is not scaled in height when drawn.
     */
    drawImage(image:CanvasImageSource, sourceX:number, sourceY:number, sourceWidth:number, sourceHeight:number,
              targetX:number, targetY:number, targetWidth:number, targetHeight:number):void;
}

File: src/modules/canvas2d/CanvasRenderingContext2D.cpp
//////////////////////////////////////////////////////////////////////////////////////
//
//  The MIT License (MIT)
//
//  Copyright (c) 2017-present, cyder.org
//  All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy of
//  this software and associated documentation files (the "Software"), to deal in the
//  Software without restriction, including without limitation the rights to use, copy,
//  modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
//  and to permit persons to whom the Software is furnished to do so, subject to the
//  following conditions:
//
//      The above copyright notice and this permission notice shall be included in all
//      copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
//  INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
//  PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
//  HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
//  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
//  SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//
//////////////////////////////////////////////////////////////////////////////////////

#include "CanvasRenderingContext2D.h"
#include <cmath>

namespace cyder {
    CanvasRenderingContext2D::CanvasRenderingContext2D(DrawingBuffer* buffer) : buffer(buffer) {
    }

    CanvasRenderingContext2D::~CanvasRenderingContext2D() {
    }

    static inline SkRect normalizeRect(const SkRect& rect) {
        return SkRect::MakeXYWH(std::min(rect.fLeft, rect.fRight),
                                std::min(rect.fTop, rect.fBottom),
                                std::max(rect.width(), -rect.width()),
                                std::max(rect.height(), -rect.height()));
    }

    void CanvasRenderingContext2D::drawImage(CanvasImageSource* image, float targetX, float targetY) {
        drawImage(image, 0, 0, image->width(), image->height(), targetX, targetY, image->width(), image->height());
    }

    void CanvasRenderingContext2D::drawImage(CanvasImageSource* image, float targetX, float targetY, float targetWidth,
                                             float targetHeight) {
        drawImage(image, 0, 0, image->width(), image->height(), targetX, targetY, targetWidth, targetHeight);
    }

    void CanvasRenderingContext2D::drawImage(CanvasImageSource* image, float sourceX, float sourceY, float sourceWidth,
                                             float sourceHeight, float targetX, float targetY, float targetWidth,
                                             float targetHeight) {
        if (!image || !image->width() || !image->height())
            return;

        if (!std::isfinite(targetX) || !std::isfinite(targetY) || !std::isfinite(targetWidth) ||
            !std::isfinite(targetHeight) || !std::isfinite(sourceX) || !std::isfinite(sourceY) ||
            !std::isfinite(sourceWidth) || !std::isfinite(sourceHeight) ||
            !targetWidth || !targetHeight || !sourceWidth || !sourceHeight) {
            return;
        }

        SkRect srcRect = normalizeRect(SkRect::MakeXYWH(sourceX, sourceY, sourceWidth, sourceHeight));
        SkRect dstRect = normalizeRect(SkRect::MakeXYWH(targetX, targetY, targetWidth, targetHeight));
        SkRect imageRect = SkRect::MakeWH(image->width(), image->height());

        if (!imageRect.contains(srcRect)) {
            // Compute the src to dst transform
            float scaleX = dstRect.width() / srcRect.width();
            float scaleY = dstRect.height() / srcRect.height();
            float offsetX = dstRect.fLeft * (scaleX - 1);
            float offsetY = dstRect.fTop * (scaleY - 1);
            srcRect.intersect(imageRect);
            // To clip the destination rectangle in the same proportion, transform the clipped src rect
            dstRect = srcRect;
            dstRect = SkRect::MakeXYWH(srcRect.x() * scaleX, srcRect.y() * scaleY,
                                       srcRect.width() * scaleX, srcRect.height() * scaleY);
            dstRect.offset(offsetX, offsetY);
        }

        if (srcRect.isEmpty()) {
            return;
        }
        image->draw(buffer->getCanvas(), dstRect, srcRect);
    }
}